Change OOB test, fix deadlocks

ramosian-glider opened the pull request.

melver commented on mm/kfence.c:
```diff
@@ -463,17 +464,27 @@ static void kfence_dump_object(int obj_index)
 
+static inline void kfence_start_report(void)
+{
+	console_lock();
```
Is this going to have other strange side-effects? Is this approved usage of 'console_lock()'?

At least a comment assuring that this is an ok thing to do would be good.

ramosian-glider replied:
I think so. At least it is in the header, is exported (i.e. available to modules) and has an unambiguous comment: https://elixir.bootlin.com/linux/latest/source/kernel/printk/printk.c#L2278

I'll add a comment to our code to clarify that.

melver replied:
What I'm worried about here is the "Acquires a lock which guarantees that the caller has exclusive access to the console system [...]", and then subsequent call to printk. What does "console system" here imply? I think it's the internal state of the console system, but says nothing about whether or not calling printk() still works.

The usage across the kernel I see is console drivers etc.

If we know that 'console_lock()' is always re-entrant, then we can probably make the claim that it's safe. If there is no guarantee that it's re-entrant, then we might eventually have a problem.

ramosian-glider replied:
Turns out it is not re-entrant, at least calling console_lock() twice freezes the task :(

ramosian-glider commented:
Reverted the console_lock() hack, moved error reporting outside the critical section instead.
This could've been made cleaner by making the object index part of `struct alloc_metadata`, but we won't need it anywhere else except reporting.
Please take another look.

melver commented on mm/kfence.c:
*obj = kfence_metadata[index]
should be simpler and safer.

ramosian-glider replied:
Ack.

melver commented on mm/kfence.c:
If index is oob, it won't return here. Is this intentional?

ramosian-glider replied:
No. Good catch!

melver commented on mm/kfence.c:
Is this function even needed?

Is it ever going to be called with index <0 || index >= KFENCE_NUM_OBJ?

Why not just do
  object = kfence_metadata[index]
at the callsites, and remove this function?

ramosian-glider replied:
I sometimes tend to over-generalize, just like here. Please stop me if you see that :)

melver commented on mm/kfence.c:
AFAIK this 0-initializes object and the memset(obj, 0, ..) isn't needed, right?

ramosian-glider replied:
I dropped copy_obj_metadata() altogether.

ramosian-glider commented:
Dropped copy_obj_data(), will merge this PR now.
Thanks for the review!
